app/src/main/java/com/github/sdpcoachme/CoachMeApplication.kt

```diff
@@ -16,24 +16,41 @@ import com.google.android.libraries.places.api.Places
 import com.google.firebase.FirebaseApp
 import com.google.firebase.database.ktx.database
 import com.google.firebase.ktx.Firebase
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.*
 
 open class CoachMeApplication : Application() {
+
+    private val USER_PREFERENCES_NAME = "coachme_preferences"
+
+    private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = USER_PREFERENCES_NAME)
+
     // For DI in testing, add reference to dependencies here
-    open lateinit var database: Database
+    open lateinit var store: CachingStore
     // We don't want to put the exact user location in the database, so it will
     // be stored here thanks to a locationProvider that will handle all location-related processes
     open lateinit var locationProvider: LocationProvider
 
     override fun onCreate() {
         super.onCreate()
         FirebaseApp.initializeApp(this)
-        database = CachingDatabase(FireDatabase(Firebase.database.reference))
+
+        store = CachingStore(FireDatabase(Firebase.database.reference), dataStore, this)
         locationProvider = FusedLocationProvider()
 
         // Initialize Places SDK
         Places.initialize(applicationContext, BuildConfig.MAPS_API_KEY)
     }
 
+    override fun onTerminate() {
+        super.onTerminate()
+        store.storeLocalData()
+    }
+
+    fun superOnTerminate() {
```

Comment:
As mentioned in CoachMeTestApplication, this naming is not a problem but a bit confusing, maybe add an explanatory comment.